components/calendar/src/chinese_based.rs

```diff
@@ -0,0 +1,346 @@
+        let new_moon_moment = Astronomical::new_moon_before(Self::midnight(moment));
+        let location = C::location(new_moon_moment.as_rata_die());
+        Location::standard_from_universal(new_moon_moment, location).as_rata_die()
+    }
+
+    /// Universal time of midnight at start of a Moment's day at the observation location
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5353-L5357
+    pub(crate) fn midnight(date: Moment) -> Moment {
+        Location::universal_from_standard(date, C::location(date.as_rata_die()))
+    }
+
+    /// Determines the fixed date of the lunar new year in the sui4 (solar year based on the winter solstice)
+    /// which contains the fixed date passed as an argument.
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5370-L5394
+    pub(crate) fn new_year_in_sui(date: RataDie) -> RataDie {
+        let prior_solstice = Self::winter_solstice_on_or_before(date); // s1
+        let following_solstice = Self::winter_solstice_on_or_before(prior_solstice + 370); // s2
+        let month_after_eleventh = Self::new_moon_on_or_after((prior_solstice + 1).as_moment()); // m12
+        let month_after_twelfth =
+            Self::new_moon_on_or_after((month_after_eleventh + 1).as_moment()); // m13
+        let next_eleventh_month = Self::new_moon_before((following_solstice + 1).as_moment()); // next-m11
+        let m12_float = month_after_eleventh.as_moment().inner();
+        let next_m11_float = next_eleventh_month.as_moment().inner();
+        let lhs_argument = libm::round((next_m11_float - m12_float) / MEAN_SYNODIC_MONTH) as i64;
+        if lhs_argument == 12
+            && (Self::no_major_solar_term(month_after_eleventh)
+                || Self::no_major_solar_term(month_after_twelfth))
+        {
+            Self::new_moon_on_or_after((month_after_twelfth + 1).as_moment())
+        } else {
+            month_after_twelfth
+        }
+    }
+
+    /// Get the moment of the nearest winter solstice on or before a given fixed date
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5359-L5368
+    pub(crate) fn winter_solstice_on_or_before(date: RataDie) -> RataDie {
+        let approx = Astronomical::estimate_prior_solar_longitude(
+            270.0,
+            Self::midnight((date + 1).as_moment()),
+        );
+        let mut iters = 0;
+        let max_iters = 367;
+        let mut day = Moment::new(libm::floor(approx.inner() - 1.0));
+        while iters < max_iters && 270.0 >= Astronomical::solar_longitude(Self::midnight(day + 1.0))
+        {
+            iters += 1;
```

Comment:
Suggestion: Make an issue for trying to optimize these searches

Reply:
I've opened this as an issue #3687